src/v2/auth.rs

```diff
@@ -35,91 +173,105 @@ impl Client {
             .await?;
 
         trace!("GET '{}' status: {:?}", r.url(), r.status());
-        let a = r
-            .headers()
+        r.headers()
             .get(reqwest::header::WWW_AUTHENTICATE)
-            .ok_or_else(|| Error::from("get_token: missing Auth header"))?;
-        let chal = String::from_utf8(a.as_bytes().to_vec())?;
+            .ok_or_else(|| {
+                Error::from(format!(
+                    "missing {:?} header",
+                    reqwest::header::WWW_AUTHENTICATE
+                ))
+            })
+            .map(ToOwned::to_owned)
+    }
 
-        let (mut auth_ep, service) = parse_hdr_bearer(chal.trim_start_matches("Bearer "))?;
+    /// Perform registry authentication and return the authenticated client.
+    ///
+    /// If Bearer authentication is used the returned client will be authorized for the requested scopes.
+    pub async fn authenticate(mut self, scopes: &[&str]) -> Result<Self> {
+        let credentials = if let Some(credentials) = self.credentials.clone() {
+            credentials
+        } else {
+            bail!("cannot authenticate without credentials");
+        };
 
-        trace!("Token provider: {}", auth_ep);
-        if let Some(sv) = service {
-            auth_ep += &format!("?service={}", sv);
-            trace!("Service identity: {}", sv);
-        }
+        self.auth = None;
 
-        Ok(auth_ep)
-    }
+        let authentication_header = self.get_www_authentication_header().await?;
+        match WwwAuthenticateHeaderContent::from_www_authentication_header(authentication_header)? {
+            WwwAuthenticateHeaderContent::Basic(_) => {
+                self.auth = Some(Auth::Basic(BasicAuth {
+                    user: credentials.0,
+                    password: Some(credentials.1),
+                }));
+            }
+            WwwAuthenticateHeaderContent::Bearer(bearer_header_content) => {
+                let auth_ep = bearer_header_content.auth_ep(scopes);
+                trace!("authenticate: token endpoint: {}", auth_ep);
 
-    /// Set the token to be used for further registry requests.
-    pub fn set_token(&mut self, token: Option<&str>) -> &Self {
-        if let Some(ref t) = token {
-            self.token = Some(t.to_string());
-        }
-        self
-    }
+                let url = reqwest::Url::parse(&auth_ep).map_err(|e| {
+                    Error::from(format!(
+                        "failed to parse url from string '{}': {}",
+                        auth_ep, e
+                    ))
+                })?;
+
+                let auth_req = match self.credentials.clone() {
```

Comment:
I think you already have a cloned `credentials` available here from the outer scope.

Reply:
On the second look I noticed that the match logic is covered earlier, so even better ;-)